core/src/main/java/com/graphhopper/routing/util/parsers/OSMMaxSpeedParser.java

```diff
@@ -39,27 +39,18 @@ public OSMMaxSpeedParser(DecimalEncodedValue carMaxSpeedEnc) {
 
     @Override
     public void handleWayTags(int edgeId, EdgeIntAccess edgeIntAccess, ReaderWay way, IntsRef relationFlags) {
-        double maxSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed"));
-        double fwdSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed:forward"));
-        if (!isValidSpeed(fwdSpeed) && isValidSpeed(maxSpeed))
-            fwdSpeed = maxSpeed;
-        double maxPossibleSpeed = MaxSpeed.UNLIMITED_SIGN_SPEED;
-        if (isValidSpeed(fwdSpeed) && fwdSpeed > maxPossibleSpeed)
-            fwdSpeed = maxPossibleSpeed;
-
-        double bwdSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed:backward"));
-        if (!isValidSpeed(bwdSpeed) && isValidSpeed(maxSpeed))
-            bwdSpeed = maxSpeed;
-        if (isValidSpeed(bwdSpeed) && bwdSpeed > maxPossibleSpeed)
-            bwdSpeed = maxPossibleSpeed;
-
-        if (!isValidSpeed(fwdSpeed))
-            fwdSpeed = UNSET_SPEED;
-        carMaxSpeedEnc.setDecimal(false, edgeId, edgeIntAccess, fwdSpeed);
+        carMaxSpeedEnc.setDecimal(false, edgeId, edgeIntAccess, getMaxSpeed(way, false));
+        carMaxSpeedEnc.setDecimal(true, edgeId, edgeIntAccess, getMaxSpeed(way, true));
+    }
 
-        if (!isValidSpeed(bwdSpeed))
-            bwdSpeed = UNSET_SPEED;
-        carMaxSpeedEnc.setDecimal(true, edgeId, edgeIntAccess, bwdSpeed);
+    private double getMaxSpeed(ReaderWay way, boolean reverse) {
+        final double maxSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed"));
+        final double directedMaxSpeed = OSMValueExtractor.stringToKmh(way.getTag(reverse ? "maxspeed:backward" : "maxspeed:forward"));
+        return isValidSpeed(directedMaxSpeed)
+                ? Math.min(directedMaxSpeed, MaxSpeed.UNLIMITED_SIGN_SPEED)
+                : isValidSpeed(maxSpeed)
+                ? Math.min(maxSpeed, MaxSpeed.UNLIMITED_SIGN_SPEED)
+                : UNSET_SPEED;
```

Comment:
`UNLIMITED_SIGN_SPEED` is basically the maximum maxspeed, right? This could even be highway-dependent. For example I saw a tagging mistake where a German residential used `maxspeed=65mph`(=`104kph`).

Anyway, `stringToKmh` already returns `UNLIMITED_SIGN_SPEED` in case of `maxspeed=none`, so this is used to distinguish between `maxspeed=none` and no max speed tag at all?

Reply:
Also it is a lot simpler than before, isn't it?